Firestore/core/src/firebase/firestore/remote/serializer.cc

```diff
@@ -551,11 +560,27 @@ std::unique_ptr<Document> Serializer::DecodeDocument(Reader* reader) const {
       case google_firestore_v1beta1_Document_name_tag:
         name = reader->ReadString();
         break;
-      case google_firestore_v1beta1_Document_fields_tag:
-        // TODO(rsgowman): Rather than overwriting, we should instead merge with
-        // the existing FieldValue (if any).
-        fields = DecodeFieldValueImpl(reader);
+      case google_firestore_v1beta1_Document_fields_tag: {
+        ObjectValue::Map::value_type fv =
+            reader->ReadNestedMessage<ObjectValue::Map::value_type>(
+                DecodeDocumentFieldsEntry);
+
+        if (!reader->status().ok()) return nullptr;
+
+        // Assumption: If we parse two entries for the map that have the same
```

Comment:
Nit: it might be better to avoid repeating the comment -- it's long, it will necessitate updating both if details ever change, and there's always a risk they might diverge. Perhaps just a brief recap and a pointer to the other comment? Strawman:
```cpp
// For duplicates, the latter overrides the former, see comment on writing object map for details
```